ludwig/backend/ray.py

```diff
@@ -0,0 +1,174 @@
+#! /usr/bin/env python
+# coding=utf-8
+# Copyright (c) 2020 Uber Technologies, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+import logging
+from collections import defaultdict
+
+import ray
+from horovod.ray import RayExecutor
+
+from ludwig.backend.base import Backend, RemoteTrainingMixin
+from ludwig.constants import NAME
+from ludwig.data.dataframe.dask import DaskEngine
+from ludwig.models.predictor import BasePredictor, RemotePredictor
+from ludwig.models.trainer import BaseTrainer, RemoteTrainer
+from ludwig.utils.tf_utils import initialize_tensorflow
+
+
+logger = logging.getLogger(__name__)
+
+
+def get_horovod_kwargs():
+    resources = [node['Resources'] for node in ray.state.nodes()]
+    use_gpu = int(ray.cluster_resources().get('GPU', 0)) > 0
+
+    # Our goal is to maximize the number of training resources we can
+    # form into a homogenous configuration. The priority is GPUs, but
```

Comment:
is it possible to support non-homogenous configurations?

Reply:
From Horovod's perspective: definitely. I think we would just need to rework the RayExecutor interface a little. Namely, you could imagine the user just saying `num_gpus=N`, then we place however many processes per host we need to in order to meet this request (so no more num_hosts or num_slots params in this mode).

Reply:
yeah, that sounds good. Let me make an issue on horovod then!